pkg/auth/providers/azuread/azuread_client.go

```diff
@@ -0,0 +1,550 @@
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := ac.httpClient.Do(req)
+	if err != nil {
+		logrus.Errorf("Received error from azure: %v", err)
+		return resp, err
+	}
+	// Check the status code
+	switch resp.StatusCode {
+	case 200:
+	case 201:
+	default:
+		var body bytes.Buffer
+		io.Copy(&body, resp.Body)
+		return resp, fmt.Errorf("Request failed, got status code: %d. Response: %s",
+			resp.StatusCode, body.Bytes())
+	}
+	return resp, nil
+}
+
+func (ac *Client) getUser(azureAccessToken string, config *v3.AzureADConfig) (Account, map[string]string, error) {
+	if azureAccessToken == "" {
+		return Account{}, nil, httperror.NewAPIError(httperror.ServerError, "No Azure Access token")
+	}
+
+	authority, graphEndpoint, _, err := ac.uriEndpointConstruction(config)
+	if err != nil {
+		logrus.Errorf("Wrong Azure Domain %v is provided", config.Domain)
+		return Account{}, nil, fmt.Errorf("Wrong Azure Domain %v is provided", config.Domain)
+	}
+	url, err := ac.getURL(authority, graphEndpoint, "USER", config, "")
+
+	resp, newAccessToken, providerInfo, err := ac.getFromAzureAD(azureAccessToken, config, url)
+	if err != nil {
+		logrus.Errorf("AzureAD getAzureADUser: GET url %v received error from Azure, err: %v", url, err)
+		return Account{}, nil, err
+	}
+	if newAccessToken != "" {
+		resp, _, providerInfo, err = ac.getFromAzureAD(newAccessToken, config, url)
+		if err != nil {
+			logrus.Errorf("AzureAD getAccessToken: GET url %v received error from azure, err: %v", url, err)
+			return Account{}, nil, err
+		}
+	}
+	defer resp.Body.Close()
+	var azureADAcct Account
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		logrus.Errorf("AzureAD getAzureADUser: error reading response, err: %v", err)
+		return Account{}, nil, err
+	}
+
+	if err := json.Unmarshal(b, &azureADAcct); err != nil {
+		logrus.Errorf("AzureAD getAzureADUser: error unmarshalling response, err: %v", err)
+		return Account{}, nil, err
+	}
+
+	return azureADAcct, providerInfo, nil
+}
+
+func (ac *Client) getGroups(azureAccessToken string, config *v3.AzureADConfig) ([]Account, map[string]string, error) {
+	if azureAccessToken == "" {
+		return []Account{}, nil, httperror.NewAPIError(httperror.ServerError, "No Azure Access token")
+	}
+	authority, graphEndpoint, _, err := ac.uriEndpointConstruction(config)
+	if err != nil {
+		logrus.Errorf("Wrong Azure Domain %v is provided", config.Domain)
+		return []Account{}, nil, fmt.Errorf("Wrong Azure Domain %v is provided", config.Domain)
+	}
+	url, err := ac.getURL(authority, graphEndpoint, "GROUP", config, "")
+
+	resp, newAccessToken, providerInfo, err := ac.getFromAzureAD(azureAccessToken, config, url)
+	if err != nil {
+		logrus.Errorf("AzureAD getAzureADGroup: GET url %v received error from Azure, err: %v", url, err)
+		return []Account{}, nil, err
+	}
+	if newAccessToken != "" {
+		resp, _, providerInfo, err = ac.getFromAzureAD(newAccessToken, config, url)
```

Comment:
why do you make the request a second time? I dont understand. you just made the same request at line 212, right?